inference-engine/include/ie_cache_manager.hpp

```diff
@@ -0,0 +1,160 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/**
+ * @brief This is a header file for the Inference Engine Cache Manager class C++ API
+ *
+ * @file ie_cache_manager.hpp
+ */
+#pragma once
+
+#include <memory>
+#include <istream>
+#include <string>
+#include <functional>
+#include "ie_api.h"
+
+namespace InferenceEngine {
+
+class Core;
+
+class CacheManagerContextImpl;
+
+/**
+ * @brief This class represents Inference Engine Caching Context.
+ *
+ * Caching context provides information to custom ICacheManager for in/out streams creation
+ * CacheManagerContext can't be created by client applications
+ */
+class INFERENCE_ENGINE_API_CLASS(CacheManagerContext) {
```

Comment:
why do we need this context? Looks like that information which it holds (device name and model cache directory) is not needed during cache entry read / write.

Reply:
Currently 'model cache directory' is used in FileStorageCacheManager
    **std::string getBlobFile(const std::string& blobHash, const CacheManagerContext& cacheManContext) const {
        return FileUtils::makePath(cacheManContext.getModelCacheDir(), blobHash + ".blob");
    }**

Reply:
Ok, but can make this directory to be a constructor argument for `FileStorageCacheManager`:
```c++
FileStorageCacheManager(const std::string & dir);
```
And in case if `CACHE_DIR` is passed to `Core::SetConfig`, it creates such cache manager internally.

Reply:
This is the idea that CacheManager will not be re-created internally when CACHE_DIR is changed, because user can set CacheManager's custom implementation
What about case when user first calls 'setCacheManager(CustomEncryptionManager)' and only than sets CACHE_DIR in ie.SetConfig? How Core will detect that cacheManager should not be re-created?

Reply:
What if we need more parameters like `ENCRYPTION_KEY`, `ENCRYPTION_SEED` and so on? From my point of view, this is not scalable to extend context for such cases and such parameters should be a part of specific derived `CacheManager` instance..